bin/notification.js

```diff
@@ -48,6 +49,28 @@ const queuePopulator = new QueuePopulator(
     zkConfig, kafkaConfig, qpConfig, httpsConfig, mConfig, rConfig, extConfigs);
 
 async.waterfall([
+    done => startProbeServer(qpConfig.probeServer, (err, probeServer) => {
+        if (err) {
+            log.error('error starting probe server', {
+                error: err,
+                method: 'NotificationQueuePopulator::startProbeServer',
+            });
+            done(err);
+            return;
```

Comment:
Nit:
```suggestion
            return done(err);
```

Reply:
I disagree on this one. Returning a callback can have weird side affects where we return a value when no value is expected.

Reply:
I am curious what's the difference between this and when we usually return the `next` function in `async.waterfall` calls? Why can we return next in the latter but not here?

Reply:
Well in the async waterfall or any other of the async chains, async handles the return value for us and likely just ignores it completely anyway making it safer. But when we return the result of the async waterfall or any other callback, we are leaking a return value that we have no clue where it is going or used. It may get assigned to a value, it may be incorrectly used as a normal response or anything else. It is misleading by saying our functions return undefined, when they don't. And the normal expectation that when we pass in a callback, we have no return value is not followed, so debugging would be harder if you didn't know we return callback values all over.

For the record, I don't think we should return next either, but since the immediate caller is going to ignore it anyway its not as bad. Or in cases where the function itself is guaranteed to return undefined anyway its not as bad. However, those two additional questions are two more steps to making sure our code does what it is supposed to do over just using two lines making it 100% clear without extra steps. I tend not to comment the suggestion at this point though just cause it is used so much in our code base that it would feel weird going against that now. But I do still disagree with it.

Reply:
Yea, async await would remove this option as we need the return values of course.